learning4e.py

```diff
@@ -453,139 +464,216 @@ def information_content(values):
     """
 
-    def decision_list_learning(examples):
+    def __init__(self, dataset):
+        self.predict.decision_list = self.decision_list_learning(set(dataset.examples))
+
+    def decision_list_learning(self, examples):
         if not examples:
             return [(True, False)]
-        t, o, examples_t = find_examples(examples)
+        t, o, examples_t = self.find_examples(examples)
         if not t:
             raise Exception
-        return [(t, o)] + decision_list_learning(examples - examples_t)
+        return [(t, o)] + self.decision_list_learning(examples - examples_t)
 
-    def find_examples(examples):
+    def find_examples(self, examples):
         """
         Find a set of examples that all have the same outcome under
         some test. Return a tuple of the test, outcome, and examples.
         """
         raise NotImplementedError
 
-    def passes(example, test):
+    def passes(self, example, test):
         """Does the example pass the test?"""
         raise NotImplementedError
 
-    def predict(example):
+    def predict(self, example):
         """Predict the outcome for the first passing test."""
-        for test, outcome in predict.decision_list:
-            if passes(example, test):
+        for test, outcome in self.predict.decision_list:
+            if self.passes(example, test):
                 return outcome
 
-    predict.decision_list = decision_list_learning(set(dataset.examples))
-
-    return predict
 
-
-def NearestNeighborLearner(dataset, k=1):
+class NearestNeighborLearner(Learner):
     """k-NearestNeighbor: the k nearest neighbors vote."""
 
-    def predict(example):
+    def __init__(self, dataset, k=1):
+        self.dataset = dataset
+        self.k = k
+
+    def predict(self, example):
         """Find the k closest items, and have them vote for the best."""
-        best = heapq.nsmallest(k, ((dataset.distance(e, example), e) for e in dataset.examples))
-        return mode(e[dataset.target] for (d, e) in best)
+        best = heapq.nsmallest(self.k, ((self.dataset.distance(e, example), e) for e in self.dataset.examples))
+        return mode(e[self.dataset.target] for (d, e) in best)
 
-    return predict
 
+class LossFunction:
+    def __init__(self, X, y):
+        self.X = X
+        self.y = y.flatten()
+
+    @staticmethod
+    def predict(X, theta):
+        return NotImplementedError
+
+    def function(self, theta):
+        return NotImplementedError
+
+    def jacobian(self, theta):
+        return NotImplementedError
 
-def LinearLearner(dataset, learning_rate=0.01, epochs=100):
-    """
-    [Section 18.6.4]
-    Linear classifier with hard threshold.
-    """
-    idx_i = dataset.inputs
-    idx_t = dataset.target
-    examples = dataset.examples
-    num_examples = len(examples)
 
-    # X transpose
-    X_col = [dataset.values[i] for i in idx_i]  # vertical columns of X
+class MeanSquaredError(LossFunction):
```

Comment:
Same here, I think we already defined loss functions in learning as actual functions, I don't see why we need to have them as classes.

Reply:
Idem. Anyway, in the next PR I will try to prepare the neural network for the use of these more explanatory classes in order to make it more modular by separating the part of backpropagation from that of optimization. As it is now, even if it works, I think it is not very intuitive in a learning sense.